### ISSUE TITLE
Standardize a label to exclude Pods from Node drain

### ISSUE THREAD

**Issue by sbueringer** [kubernetes/kubernetes#127247]
### What would you like to be added?

I would like to have a standardized annotation to exclude Pods from Node drains.

This annotation should be implement by kubectl drain, but other tools like e.g. Cluster API would also be able to implement this annotation.

### Context

Today there is no standard way to exclude Pods from `kubectl drain`.

There is a set of Pods that are already excluded today:
* Pods belonging to an existing DaemonSet (orphaned DaemonSet Pods are evicted as well)
* Mirror Pods, i.e. Pods with the `kubernetes.io/config.mirror` annotation (usually static Pods managed by kubelet, like kube-apiserver)

So basically the best available option to exclude a Pod from drain would be adding the `kubernetes.io/config.mirror` annotation. But this is pretty bad (and also has unknown side effects).

Would it be possible to support an annotation like e.g. `node.kubernetes.io/exclude-from-drain`

**Comment by sbueringer** [comment]
cc @enxebre @fabriziopandini @chrischdi (in case you want to add more color)

**Comment by sftim** [comment]
I recommend making this a label, not an annotation. My motivation is that you might want to select on that label. For example, a controller might want to watch for Pods that are bound to a Node and _not_ marked as exempt from draining. 

Maybe one day watch can be optimized to return minimal metadata - potentially just `.metadata.uid` (even ignoring namespace and name, which aren't relevant here).

As soon as the watch produces an empty list, the controller adds a new taint and triggers Node shutdown.

**Comment by sbueringer** [comment]
Makes sense

/retitle Standardize a label to exclude Pods from Node drain

**Comment by spacemule** [comment]
I think this is necessary, but it also touches on the issue of dependent pods. e.g. I have a database pod with storage on ceph. If I want to drain both the database and the OSD pods and they terminate in the wrong order, the database pods get stuck. If the label was set to something like:

```
pod.kubernetes.io/drain-priority = {int}
```

higher values could be terminated before any lower values, and 0 or -1 could be set to never drain. This would solve the issue of pods that shouldn't be drained while allowing flexibility for other situations.

**Comment by sftim** [comment]
/sig architecture
(in that this spans several SIGs)

**Comment by dims** [comment]
@sftim would love folks interested to come up with options we can then debate on.
